tests/unit/domain/validation/test_validation_new_features.py

```
@@ -49,6 +49,16 @@ def test_template_lints_vt2_vt3_vt4() -> None:
     pipeline = load_pipeline_blueprint_from_yaml(yaml_text)
     report = pipeline.validate_graph()
     rules = {w.rule_id for w in report.warnings}
+
+    # Debug: Print all warnings found (helps debug CI issues)
+    if "V-T4" not in rules:
+        import os
+        if os.getenv("CI") == "true":
+            print(f"DEBUG: Expected V-T4 but only found rules: {rules}")
+            print(f"DEBUG: All warnings:")
```

Ruff (0.14.5)

f-string without any placeholders

Remove extraneous `f` prefix

(F541)